File: app/src/main/java/com/example/megaport/mynews/Controllers/Activities/MainActivity.java
package com.example.megaport.mynews.Controllers.Activities;

import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.design.widget.NavigationView;
import android.support.design.widget.TabLayout;
import android.support.v4.view.GravityCompat;
import android.support.v4.view.ViewPager;
import android.support.v4.widget.DrawerLayout;
import android.support.v7.app.ActionBarDrawerToggle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.Menu;
import android.view.MenuItem;
import com.example.megaport.mynews.Controllers.Fragments.MainFragment;
import com.example.megaport.mynews.Controllers.Fragments.fragmentPagerAdapter;
import com.example.megaport.mynews.R;

public class MainActivity extends AppCompatActivity implements NavigationView.OnNavigationItemSelectedListener {

    //FOR DESIGN
    private Toolbar toolbar;
    private DrawerLayout drawerLayout;

    // FOR RECYCLERVIEW
    private MainFragment mainFragment = (MainFragment) getSupportFragmentManager().findFragmentById(R.id.activity_main_frame_layout);
    public static int tabPos = 0;
    private ViewPager mViewPager;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        this.configureToolBar();

        this.configureDrawerLayout();

        this.configureNavigationView();


        if(savedInstanceState == null) {
            this.configureAndShowMainFragment();
        }
        else {
            mainFragment = (MainFragment) getSupportFragmentManager().findFragmentById(R.id.activity_main_frame_layout);
        }

        // Get the ViewPager and set it's PagerAdapter so that it can display items
        mViewPager = findViewById(R.id.viewpager);
        mViewPager.setAdapter(new fragmentPagerAdapter(getSupportFragmentManager()
        ));

        // Give the TabLayout the ViewPager
        TabLayout tabLayout = findViewById(R.id.sliding_tabs);
        tabLayout.setupWithViewPager(mViewPager);
        tabLayout.addOnTabSelectedListener(new TabLayout.ViewPagerOnTabSelectedListener(mViewPager) {
            @Override
            public void onTabSelected(TabLayout.Tab tab) {
                super.onTabSelected(tab);
                mViewPager.setCurrentItem(tab.getPosition());
                tabPos = tab.getPosition();
                mainFragment.executeHttpRequestWithRetrofit();
            }

            @Override
            public void onTabReselected(TabLayout.Tab tab) {
                mainFragment.executeHttpRequestWithRetrofit();
            }
        });
    }

    // --------
    // TOOLBAR
    // --------

    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.activity_main_menu_toolbar, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()){
            case R.id.activity_main_toolbar_search:
                Intent intentSearch = new Intent(this, SearchActivity.class);
                startActivity(intentSearch);
                break;
            case R.id.activity_main_toolbar_help:
                Intent intentHelp = new Intent( this, HelpActivity.class );
                startActivity(intentHelp);
                break;
            case R.id.activity_main_toolbar_about:
                Intent intentAbout = new Intent(this, AboutActivity.class);
                startActivity(intentAbout);
                break;
            default:
                break;
        }

        return true;
    }

    // -------------

    @Override
    public void onBackPressed() {
        // 5 - Handle back click to close menu
        if (this.drawerLayout.isDrawerOpen(GravityCompat.START)) {
            this.drawerLayout.closeDrawer(GravityCompat.START);
        } else {
            super.onBackPressed();
        }
    }

    // -------------------
    // NAVIGATION DRAWER
    // -------------------

    @Override
    public boolean onNavigationItemSelected(@NonNull MenuItem item) {

        // 4 - Handle Navigation Item Click
        int id = item.getItemId();

        switch (id){
            case R.id.activity_main_drawer_top_stories :
                mViewPager.setCurrentItem( 0 );
                break;
            case R.id.activity_main_drawer_most_popular:
                mViewPager.setCurrentItem( 1 );
                break;
            case R.id.activity_main_drawer_notification:
                Intent intentNotification = new Intent(this, NotificationsActivity.class);
                startActivity(intentNotification);
                break;
            case R.id.activity_main_drawer_help:
                Intent intentHelp = new Intent( this, HelpActivity.class );
                startActivity(intentHelp);
                break;
            case R.id.activity_main_drawer_about:
                Intent intentAbout = new Intent(this, AboutActivity.class);
                startActivity(intentAbout);
                break;
            default:
                break;
        }

        this.drawerLayout.closeDrawer(GravityCompat.START);

        return true;
    }

    // ---------------------
    // CONFIGURATION
    // ---------------------

    // 1 - Configure Toolbar
    private void configureToolBar(){
        this.toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
    }

    // 2 - Configure Drawer Layout
    private void configureDrawerLayout(){
        this.drawerLayout = findViewById(R.id.activity_main_drawer_layout);
        ActionBarDrawerToggle toggle = new ActionBarDrawerToggle(this, drawerLayout, toolbar, R.string.navigation_drawer_open, R.string.navigation_drawer_close);
        drawerLayout.addDrawerListener(toggle);
        toggle.syncState();
    }

    // 3 - Configure NavigationView
    private void configureNavigationView(){
        NavigationView navigationView = findViewById( R.id.activity_main_nav_view );
        navigationView.setNavigationItemSelectedListener(this);
    }

    private void configureAndShowMainFragment(){
        if (mainFragment == null) {
            mainFragment = new MainFragment();
            getSupportFragmentManager().beginTransaction()
                    .add(R.id.activity_main_frame_layout, mainFragment)
                    .commit();
        }
    }

}

File: app/src/main/java/com/example/megaport/mynews/controllers/activities/MainActivity.kt
package com.example.megaport.mynews.controllers.activities

import android.content.Intent
import android.os.Bundle
import com.google.android.material.navigation.NavigationView
import com.google.android.material.tabs.TabLayout
import androidx.core.view.GravityCompat
import androidx.appcompat.app.ActionBarDrawerToggle
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import android.view.Menu
import android.view.MenuItem
import com.example.megaport.mynews.controllers.fragments.FragmentPagerAdapter
import com.example.megaport.mynews.R
import com.example.megaport.mynews.controllers.utils.openActivity
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity : AppCompatActivity(), NavigationView.OnNavigationItemSelectedListener {

    private var drawerLayout: androidx.drawerlayout.widget.DrawerLayout? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        configureToolBar()

        configureDrawerLayout()

        configureNavigationView()

        // Get the ViewPager and set it's PagerAdapter so that it can display items
        viewpager.adapter = FragmentPagerAdapter(supportFragmentManager)

        // Give the TabLayout the ViewPager
        val tabLayout = findViewById<TabLayout>(R.id.sliding_tabs)
        tabLayout.setupWithViewPager(viewpager)
    }

    // --------
    // TOOLBAR
    // --------

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.activity_main_menu_toolbar, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.activity_main_toolbar_search -> {
                openActivity<SearchActivity>()
            }
            R.id.activity_main_toolbar_help -> {
                openActivity<HelpActivity>()
            }
            R.id.activity_main_toolbar_about -> {
                openActivity<AboutActivity>()
            }
            else -> {
                throw IllegalArgumentException("Invalide item")
            }
        }
        return true
    }

    // -------------

    override fun onBackPressed() {
        // 5 - Handle back click to close menu
        if (this.drawerLayout!!.isDrawerOpen(GravityCompat.START)) {
            this.drawerLayout!!.closeDrawer(GravityCompat.START)
        } else {
            super.onBackPressed()
        }
    }

    // -------------------
    // NAVIGATION DRAWER
    // -------------------

    override fun onNavigationItemSelected(item: MenuItem): Boolean {

        val id = item.itemId

        when (id) {

            R.id.activity_main_drawer_arts -> {
                navigateToArticleList("arts")
            }
            R.id.activity_main_drawer_technology-> {
                navigateToArticleList("technology")
            }
            R.id.activity_main_drawer_automobiles-> {
                navigateToArticleList("automobiles")
            }
            R.id.activity_main_drawer_food-> {
                navigateToArticleList("food")
            }
            R.id.activity_main_drawer_movies-> {
                navigateToArticleList("movies")
            }
            R.id.activity_main_drawer_sports->{
                navigateToArticleList("sports")
            }
            R.id.activity_main_drawer_science-> {
                navigateToArticleList("science")
            }
            R.id.activity_main_drawer_books-> {
                navigateToArticleList("books")
            }
            R.id.activity_main_drawer_world-> {
                navigateToArticleList("world")
            }
            R.id.activity_main_drawer_magazine->{
                navigateToArticleList("magazine")
            }

            // ---------------------------------------------

            R.id.activity_main_drawer_notification -> {
                openActivity<NotificationsActivity>()
            }
            R.id.activity_main_drawer_help -> {
                openActivity<HelpActivity>()
            }
            R.id.activity_main_drawer_about -> {
                openActivity<AboutActivity>()
            }
            else -> {
            }
        }

        this.drawerLayout!!.closeDrawer(GravityCompat.START)

        return true
    }

    private fun navigateToArticleList(section:String){
        val intent = Intent(this, ArticleListActivity::class.java)
        intent.putExtra(SECTION,section)
        startActivity(intent)
    }

    // ---------------------
    // CONFIGURATION
    // ---------------------

    private fun configureToolBar() {
        val toolbar = findViewById<Toolbar>(R.id.toolbar)
        setSupportActionBar(toolbar)
    }

    // 2 - Configure Drawer Layout
    private fun configureDrawerLayout() {
        this.drawerLayout = findViewById(R.id.activity_main_drawer_layout)
        val toggle = ActionBarDrawerToggle(this, drawerLayout, toolbar, R.string.navigation_drawer_open, R.string.navigation_drawer_close)
        drawerLayout?.addDrawerListener(toggle)
        toggle.syncState()
    }

    // 3 - Configure NavigationView
    private fun configureNavigationView() {
        val navigationView = findViewById<NavigationView>(R.id.activity_main_nav_view)
        navigationView.setNavigationItemSelectedListener(this)
    }

    companion object{
        private const val SECTION = "section"

    }

}